lib/glific/access_control.ex

```diff
@@ -0,0 +1,285 @@
+
+      iex> list_permissions()
+      [%Permission{}, ...]
+
+  """
+  @spec list_permissions :: [Permission.t()]
+  def list_permissions do
+    Repo.all(Permission)
+  end
+
+  @doc """
+  Gets a single permission.
+
+  Raises `Ecto.NoResultsError` if the Permission does not exist.
+
+  ## Examples
+
+      iex> get_permission!(123)
+      %Permission{}
+
+      iex> get_permission!(456)
+      ** (Ecto.NoResultsError)
+
+  """
+  @spec get_permission!(integer) :: Permission.t()
+  def get_permission!(id), do: Repo.get!(Permission, id)
+
+  @doc """
+  Creates a permission.
+
+  ## Examples
+
+      iex> create_permission(%{field: value})
+      {:ok, %Permission{}}
+
+      iex> create_permission(%{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+
+  """
+  @spec create_permission(map()) :: {:ok, Permission.t()} | {:error, Ecto.Changeset.t()}
+  def create_permission(attrs \\ %{}) do
+    %Permission{}
+    |> Permission.changeset(attrs)
+    |> Repo.insert()
+  end
+
+  @doc """
+  Updates a permission.
+
+  ## Examples
+
+      iex> update_permission(permission, %{field: new_value})
+      {:ok, %Permission{}}
+
+      iex> update_permission(permission, %{field: bad_value})
+      {:error, %Ecto.Changeset{}}
+
+  """
+  @spec update_permission(Permission.t(), map()) ::
+          {:ok, Permission.t()} | {:error, Ecto.Changeset.t()}
+  def update_permission(%Permission{} = permission, attrs) do
+    permission
+    |> Permission.changeset(attrs)
+    |> Repo.update()
+  end
+
+  @doc """
+  Deletes a permission.
+
+  ## Examples
+
+      iex> delete_permission(permission)
+      {:ok, %Permission{}}
+
+      iex> delete_permission(permission)
+      {:error, %Ecto.Changeset{}}
+
+  """
+  @spec delete_permission(Permission.t()) :: {:ok, Permission.t()} | {:error, Ecto.Changeset.t()}
+  def delete_permission(%Permission{} = permission) do
+    Repo.delete(permission)
+  end
+
+  @doc """
+  Common function to filtering entity objects based on user role, fun_with_flag flag and entity type
+  """
+  @spec check_access(Ecto.Query.t(), atom()) :: Ecto.Query.t()
+  def check_access(entity_list, entity_type) do
+    user = Repo.get_current_user() |> Repo.preload([:access_roles])
+
+    if check_roles_and_permission_toggle?(user.organization_id) and
+         is_organization_role?(user),
+       do: do_check_access(entity_list, entity_type, user),
+       else: entity_list
+  end
+
+  @spec is_organization_role?(User.t() | nil) :: boolean()
+  defp is_organization_role?(nil), do: false
+
+  defp is_organization_role?(%{access_roles: access_roles} = _user) do
+    !Enum.any?(access_roles, fn access_role ->
+      access_role.label in ["Admin", "Manager", "Staff"]
+    end)
+  end
+
+  @doc """
+  check fun_with_flag toggle for an organization and returns boolean value
+  """
+  @spec check_roles_and_permission_toggle?(non_neg_integer()) :: boolean()
+  def check_roles_and_permission_toggle?(organization_id) do
+    FunWithFlags.enabled?(
```

Comment:
Can we move this to org services also? 

Then we will be able to use it like organization.services['roles_and_permission'] 

which will help frontend also

Reply:
@pankaj-ag done